lib/northbound.h

```diff
@@ -0,0 +1,777 @@
+	 * has been deleted.
+	 *
+	 * The callback is supposed to delete the entire configuration object,
+	 * including its children when they exist.
+	 *
+	 * event
+	 *    The transaction phase. Refer to the documentation comments of
+	 *    nb_event for more details.
+	 *
+	 * dnode
+	 *    libyang data node that is being deleted.
+	 *
+	 * Returns:
+	 *    - NB_OK on success.
+	 *    - NB_ERR_VALIDATION when a validation error occurred.
+	 *    - NB_ERR_INCONSISTENCY when an inconsistency was detected.
+	 *    - NB_ERR for other errors.
+	 */
+	int (*delete)(enum nb_event event, const struct lyd_node *dnode);
+
+	/*
+	 * Configuration callback.
+	 *
+	 * A list entry or leaf-list entry has been moved. Only applicable when
+	 * the "ordered-by user" statement is present.
+	 *
+	 * event
+	 *    The transaction phase. Refer to the documentation comments of
+	 *    nb_event for more details.
+	 *
+	 * dnode
+	 *    libyang data node that is being moved.
+	 *
+	 * Returns:
+	 *    - NB_OK on success.
+	 *    - NB_ERR_VALIDATION when a validation error occurred.
+	 *    - NB_ERR_INCONSISTENCY when an inconsistency was detected.
+	 *    - NB_ERR for other errors.
+	 */
+	int (*move)(enum nb_event event, const struct lyd_node *dnode);
+
+	/*
+	 * Optional configuration callback.
+	 *
+	 * The 'apply_finish' callbacks are called after all other callbacks
+	 * during the apply phase (NB_EV_APPLY). These callbacks are called only
+	 * under one of the following two cases:
+	 * - The data node has been created or modified (but not deleted);
+	 * - Any change was made within the descendants of the data node (e.g. a
+	 *   child leaf was modified, created or deleted).
+	 *
+	 * In the second case above, the 'apply_finish' callback is called only
+	 * once even if multiple changes occurred within the descendants of the
+	 * data node.
+	 *
+	 * dnode
+	 *    libyang data node associated with the 'apply_finish' callback.
+	 */
+	void (*apply_finish)(const struct lyd_node *dnode);
+
+	/*
+	 * Operational data callback.
+	 *
+	 * The callback function should return the value of a specific leaf or
+	 * inform if a typeless value (presence containers or leafs of type
+	 * empty) exists or not.
+	 *
+	 * xpath
+	 *    YANG data path of the data we want to get.
+	 *
+	 * list_entry
+	 *    Pointer to list entry.
+	 *
+	 * Returns:
+	 *    Pointer to newly created yang_data structure, or NULL to indicate
+	 *    the absence of data.
+	 */
+	struct yang_data *(*get_elem)(const char *xpath,
+				      const void *list_entry);
+
+	/*
+	 * Operational data callback for YANG lists.
+	 *
+	 * The callback function should return the next entry in the list. The
+	 * 'list_entry' parameter will be NULL on the first invocation.
+	 *
+	 * list_entry
+	 *    Pointer to list entry.
+	 *
+	 * Returns:
+	 *    Pointer to the next entry in the list, or NULL to signal that the
+	 *    end of the list was reached.
+	 */
+	const void *(*get_next)(const void *list_entry);
```

Comment:
As we discussed previously, I think these methods should have an `xpath` parameter to uniquely identify the owner of the list whenever it is not global (e.g. a member of a specific IS-IS area, BGP peer etc.).